helix-term/src/ui/editor.rs

```diff
@@ -44,6 +44,124 @@ impl Default for EditorView {
     }
 }
 
+struct Merge<I> {
+    iter: I,
+    spans: Box<dyn Iterator<Item = (usize, std::ops::Range<usize>)>>,
+
+    next_event: Option<HighlightEvent>,
+    next_span: Option<(usize, std::ops::Range<usize>)>,
+
+    queue: Vec<HighlightEvent>,
+}
+
+fn merge<I: Iterator<Item = HighlightEvent>>(
+    iter: I,
+    spans: Vec<(usize, std::ops::Range<usize>)>,
+) -> impl Iterator<Item = HighlightEvent> {
```

Comment:
Nit: is there a reason the return type needs to be abstracted like this?  If I'm reading the code right, it looks like it always returns a `Merge`, so I think that would be clearer.

Reply:
Yeah I simplified that now. Was returning `impl Iterator` to avoid naming a specific type, but `std` seems to always export the iterator structs so I did the same.